keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,1113 @@
+```
+
+Apply the same manifest with `--server-side` flag, as per server-side-apply migration instructions:
+```shell
+cat <<EOF | kubectl apply --server-side -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+  legacy: unused
+EOF
+```
+```console
+configmap/test serverside-applied
+```
+
+Remove one of the values from the configmap, we choose `legacy` key, and apply
+again using server-side apply. You would expect that the field is now removed
+from the object, but it is not:
+```shell
+cat <<EOF | kubectl apply --server-side -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+EOF
+```
+```console
+configmap/test serverside-applied
+```
+
+Check the configmap, expecting the legacy value to be removed, and be surprised that it remains:
+```shell
+kubectl get configmap -o yaml test
+```
+```console
+apiVersion: v1
+kind: ConfigMap
+data:
+  key: value
+  legacy: unused
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","data":{"key":"value"},"kind":"ConfigMap","metadata":{"name":"test","namespace":"default"}}
+  name: test
+  namespace: default
+```
+
+This KEP is intended to fix this problem by providing users with a path to completely
+migrate their managed fields to be used with server-side-apply.
+
+### Goals
+
+This bug prevents users of server-side-apply, which is in GA, from deleting any fields of their objects unless they revert back to using client-side-apply. It especially poses huge
+challenges to cluster operators seeking to upgrade a large collection of resources to server-side-apply. Without an automatic method of preparing objects for use with SSA,
+adoption for server-side-apply is blocked for many users.
+
+1. The primary goal of this KEP is to provide a pathway for users blocked by this issue to adopt SSA.
+2. Provide an idempotent migration to convert large collections of resources to server-side apply, especially for automatic systems
+3. Keep the number of requests to the apiserver limited on the nominal server-side apply path (no additional writes)
+
+
+### Non-Goals
+
+1. Provide a server-side fix to this problem (see alternatives considered)
+2. Deprecate the `kubectl/last-applied-configuration` annotation
+
+## Proposal
+
+Provide a transparent migration path of client-side-apply field ownership to
+being owned by server-side-apply.
+
+```shell
+kubectl apply -f <path/to/file> --server-side
+```
+
+Today when using server-side-apply, kubectl unconditionally does an initial fetch
+of the object. This is an acknowledged anti-pattern of SSA, and will likely be changed
+in the future. Since it is pre-existing behavior that does not do much harm, the SIG has
+come to consensus that the behavior can remain for a few more releases until client-side-apply
+is out of vogue, especially since SSA is now GA.
+
+The initially fetched object will be tested to check if the migration should be
```

Comment:
Since the migration to SSA is idempotent, this is determined by seeing if the migration was a no-op. If there is any work to do, then it will be sent as an UPDATE to the server before continuing with the apply as normal